src/xrpld/app/tx/detail/OnChainCredentials.cpp

```diff
@@ -0,0 +1,454 @@
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/detail/OnChainCredentials.h>
+
+#include <xrpld/core/TimeKeeper.h>
+#include <xrpld/ledger/ApplyView.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/basics/Log.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/TxFlags.h>
+#include <xrpl/protocol/st.h>
+
+#include <chrono>
+
+namespace ripple {
+
+/*
+    OnChainCredentials
+    ======
+
+   A verifiable credentials (VC
+   https://en.wikipedia.org/wiki/Verifiable_credentials), as defined by the W3C
+   specification (https://www.w3.org/TR/vc-data-model-2.0/), is a
+   secure and tamper-evident way to represent information about a subject, such
+   as an individual, organization, or even an IoT device. These credentials are
+   issued by a trusted entity and can be verified by third parties without
+   directly involving the issuer at all.
+*/
+
+//------------------------------------------------------------------------------
+
+bool
+credentialCheckExpired(
+    std::shared_ptr<SLE const> const& sle,
+    NetClock::time_point const& closed)
+{
+    std::uint32_t const exp =
+        sle->isFieldPresent(sfExpiration) ? sle->getFieldU32(sfExpiration) : 0;
+    std::uint32_t const now = closed.time_since_epoch().count();
+    return now > exp;
+}
+
+// special check for deletion
+static bool
+credentialCheckNotExpired(
+    std::shared_ptr<SLE const> const& sle,
+    NetClock::time_point const& closed)
+{
+    if (!sle->isFieldPresent(sfExpiration))
+        return false;
+
+    std::uint32_t const exp = sle->getFieldU32(sfExpiration);
+    std::uint32_t const now = closed.time_since_epoch().count();
+
+    return now <= exp;
+}
+
+// ------- CREATE --------------------------
+
+NotTEC
+CredentialCreate::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureOnChainCredentials))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+    auto& j = ctx.j;
+
+    auto const optSubj = tx[~sfSubject];
+    auto const optIss = tx[~sfIssuer];
+    auto const optSig = tx[~sfSignature];
+
+    if ((optSubj.has_value() == optIss.has_value()) ||
+        (optIss.has_value() != optSig.has_value()))
+    {
+        // Either both fields are present or neither field is present.  In
+        // either case the transaction is malformed.
+        JLOG(j.trace())
+            << "Malformed transaction: "
+               "Invalid Subject, Issuer and Signature fields combination.";
+        return temMALFORMED;
+    }
+
+    // Make sure that the passed account is valid.
+    AccountID const target{optSubj ? *optSubj : *optIss};
+    if (!target)
+    {
+        JLOG(j.trace()) << "Malformed transaction: Subject or Issuer "
+                           "field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
```

Comment:
This check should already have been performed by the transaction validation.

Reply:
The same is in `CredentialDelete::preflight` and `CredentialAccept::preflight`.

Reply:
There are tests that fire these checks. Since these fields are optional, maybe transaction validation is not so strict.